### PR TITLE
Integrate sidecar shim image

### LINKED ISSUE #11161: Make `image` field optional in the `HookSidecar` struct
Presently, it's mandatory for a hook sidecar user to provide a container image in the annotations when using the feature. Making this field optional by using an image with digest in virt-operator, much like how it's done in [this commit](https://github.com/kubevirt/kubevirt/pull/9177/commits/88bee3a38e1ab0bbbc3bc5f6fc7605cffc33be4a), would be helpful.

Using an image with digest that is based on some crypto function helps ensure that an image is not tempered with, which could plausibly be the case when using plain tags. This is especially useful in case of disconnected environment where images need to be pulled onto internal registry in advance.

### PR DESCRIPTION
**Pull request by dharmit**
### What this PR does
Before this PR: `image` field in the `hooks.kubevirt.io/hookSidecars` annotation in [this example](https://github.com/kubevirt/kubevirt/blob/main/examples/vmi-with-sidecar-hook-configmap.yaml) was mandatory.

After this PR: aforementioned field will be optional.

Fixes #11161 

### Why we need it and why it was done in this way
The following tradeoffs were made: N/A

The following alternatives were considered: N/A

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Make 'image' field in hook sidecar annotation optional.
```

### CODE REVIEW

**Comment by dharmit** [comment]
@alicefr @victortoso @vasiliy-ul can you folks help with review? Pinging you folks since I have noticed you being active on sidecar hook side of the things. :)

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -22,7 +22,9 @@ package services
 import (
 	"context"
 	"fmt"
+	operatorutil "kubevirt.io/kubevirt/pkg/virt-operator/util"
```
Comment (vasiliy-ul) [inline pkg/virt-controller/services/template.go:25]: nit: can we move this to the group with local imports?

**Review by vasiliy-ul — COMMENTED** [review]
Overall, this looks okay. One thing to consider: we already have some tests for sidecars in `tests/vmi_hook_sidecar_test.go`. But they all explicitly set the image. Perhaps we shall consider changing that, so at least a couple of tests run with empty image. WDYT?

**Comment by vasiliy-ul** [comment]
There are some formatting issues.

**Comment by vasiliy-ul** [comment]
Also, unit-test failure seems caused by this change.

File: pkg/virt-controller/services/template.go (hunk 1 above)
Comment (dharmit) [reply pkg/virt-controller/services/template.go]: Ack. Done.

**Comment by dharmit** [comment]
> Perhaps we shall consider changing that, so at least a couple of tests run with an empty image. WDYT?

I agree. I have added one test. If you have a suggestion for more tests in the same file, let me know. I'm open to add more.

**Comment by dharmit** [comment]
> There are some formatting issues.

Can you point? I found that running `make test` modified the bazel file. Is that the one you're talking about?

> Also, unit-test failure seems caused by this change.

Thanks. Fixed.

**Comment by vasiliy-ul** [comment]
> Can you point? I found that running make test modified the bazel file. Is that the one you're talking about?

@dharmit, you can see the formatting issue in the details for the `pull-kubevirt-build` job: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11272/pull-kubevirt-build/1764935860379193344
```diff
diff -u ./pkg/hooks/hooks.go.orig ./pkg/hooks/hooks.go
--- ./pkg/hooks/hooks.go.orig	2024-03-05 08:50:01.812356865 +0000
+++ ./pkg/hooks/hooks.go	2024-03-05 08:50:01.812356865 +0000
@@ -21,6 +21,7 @@
 
 import (
 	"encoding/json"
+
 	k8sv1 "k8s.io/api/core/v1"
 
 	v1 "kubevirt.io/api/core/v1"
diff -u ./pkg/virt-controller/services/template.go.orig ./pkg/virt-controller/services/template.go
--- ./pkg/virt-controller/services/template.go.orig	2024-03-05 08:50:03.168358416 +0000
+++ ./pkg/virt-controller/services/template.go	2024-03-05 08:50:03.168358416 +0000
@@ -22,12 +22,13 @@
 import (
 	"context"
 	"fmt"
-	operatorutil "kubevirt.io/kubevirt/pkg/virt-operator/util"
 	"math/rand"
 	"os"
 	"strconv"
 	"strings"
 
+	operatorutil "kubevirt.io/kubevirt/pkg/virt-operator/util"
+
 	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
 	k8sv1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
```
Basically, you just need to build your changes locally and the autoformatter will adjust everything as needed. Then simply add and commit the changes.

One of the formatting issue is actually about https://github.com/kubevirt/kubevirt/pull/11272#discussion_r1512365224, which you probably fixed already.

File: tests/vmi_hook_sidecar_test.go [hunk 2]
```diff
@@ -303,10 +303,22 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
...
+				vmi.ObjectMeta.Annotations = RenderSidecarWithConfigMapPlusImage(hooksv1alpha2.Version, cm.Name)
+				vmi = tests.RunVMIAndExpectLaunch(vmi, 360)
+				domainXml, err := tests.GetRunningVirtualMachineInstanceDomainXML(virtClient, vmi)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(domainXml).Should(ContainSubstring("<sysinfo type='smbios'>"))
+				Expect(domainXml).Should(ContainSubstring("<smbios mode='sysinfo'/>"))
+				Expect(domainXml).Should(ContainSubstring("<entry name='manufacturer'>Radical Edward</entry>"))
+			})
+
+			It("should update domain XML with SM BIOS properties when sidecar image is not specified", func() {
```
Comment (vasiliy-ul) [inline tests/vmi_hook_sidecar_test.go:318]: The two tests look almost identical. Could we parametrize them as needed and use `DescribeTable` to avoid duplicated code?

File: hack/generate.sh [hunk 3]
```diff
@@ -200,7 +202,28 @@ prhelper_rawsha=$(getRawShasum ".PrHelperSha")
...
+    --exportProxySha="$virtexportproxy_rawsha" \
+    --exportServerSha="$virtexportserver_rawsha" \
+    --gsSha="$gs_rawsha" \
+    --prHelperSha="$prhelper_rawsha" \
+    --kubevirtLogo={{.KubeVirtLogo}} \
+    --csvVersion="$_fake_csv_version" \
+    --replacesCsvVersion="$_fake_replaces_csv_version" \
+    --csvCreatedAtTimestamp={{.CreatedAt}} \
+    --kubeVirtVersion={{.DockerTag}} \
+    --sidecarShimSha="$sidecar_shim_rawsha" \
```
Comment (vasiliy-ul) [inline hack/generate.sh:224]: nit: maybe sot this alphabetically?

File: tests/vmi_hook_sidecar_test.go [hunk 4]
```diff
@@ -362,13 +374,20 @@ func RenderInvalidSMBiosSidecar() map[string]string {
 	}
 }
 
-func RenderSidecarWithConfigMap(version, name string) map[string]string {
+func RenderSidecarWithConfigMapPlusImage(version, name string) map[string]string {
 	return map[string]string{
 		"hooks.kubevirt.io/hookSidecars": fmt.Sprintf(`[{"args": ["--version", "%s"], "image":"%s/%s:%s", "configMap": {"name": "%s","key": "%s", "hookPath": "/usr/bin/onDefineDomain"}}]`,
 			version, flags.KubeVirtUtilityRepoPrefix, sidecarShimImage, flags.KubeVirtUtilityVersionTag, name, configMapKey),
 	}
 }
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply tests/vmi_hook_sidecar_test.go:382]: > quay.io/kubevirt/sidecar-shim:latest

This does not look right to me. When you build from sources and test with local cluster using `make cluster-up && make cluster-sync`, `sidecar-shim` image should be pushed to a local registry and the image path should be rendered as `registry:5000/kubevirt/sidecar-shim:devel` (or smth similar). How do you run the tests? 

You can get `quay.io/kubevirt/sidecar-shim:latest` if you run the test suite manually without providing the needed flags. Is it the case?

[the remark this answers is not in the document]
Comment (dharmit) [reply tests/vmi_hook_sidecar_test.go:382]: > How do you run the tests?
> 
> You can get `quay.io/kubevirt/sidecar-shim:latest` if you run the test suite manually without providing the needed flags. Is it the case?

Looks like this is the case. I was running through the IDE. Running `make functest` passes without any failure. 

Apologies for the noise.

File: tests/vmi_hook_sidecar_test.go (hunk 2 above)
Comment (dharmit) [reply tests/vmi_hook_sidecar_test.go]: Done.

File: hack/generate.sh (hunk 3 above)
Comment (dharmit) [reply hack/generate.sh]: Done.

**Comment by dharmit** [comment]
> /test pull-kubevirt-e2e-k8s-1.27-sig-compute
> /test pull-kubevirt-e2e-k8s-1.28-sig-compute
> /test pull-kubevirt-e2e-k8s-1.29-sig-compute

All three failures have below logs:
```
13:47:38: + /home/prow/go/src/github.com/kubevirt/kubevirt/_out/cmd/dump/dump --kubeconfig=
13:47:39: W0307 13:47:39.396152   17558 client_config.go:618] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
13:47:39: W0307 13:47:39.396218   17558 client_config.go:623] error creating inClusterConfig, falling back to default config: open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
13:47:39: failed to get client config: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable 
```
This doesn't look related to the changes I have made.

**Comment by dharmit** [comment]
> This doesn't look related to the changes I have made.

@vasiliy-ul WDYT? Can you PTAL?

**Comment by alicefr** [comment]
@dharmit the code looks good. 
Just a suggestion, I would split the commit in 2 part, one with the rearrangements of the variables and a second wit the new variables for the sidecar sha image. It is just easier to check the new variable.

**Comment by dharmit** [comment]
@alicefr are you referring to 957aaa1? I made a mistake in that commit of adding the sorting of CLI flags in that commit instead of the earlier commit.

**Comment by alicefr** [comment]
here https://github.com/kubevirt/kubevirt/pull/11272/commits/cf98f5ffe4df7e560242c707892efbaa0fafd513 you are changing the variables order and also adding the new image. It is a bit confusing. If you need to sort the variable I'd do in a previous commit, so then we can clearly identify the new image.

**Comment by vasiliy-ul** [comment]
Lets give it another try

/test pull-kubevirt-e2e-k8s-1.27-sig-compute
/test pull-kubevirt-e2e-k8s-1.28-sig-compute
/test pull-kubevirt-e2e-k8s-1.29-sig-compute

**Review by vasiliy-ul — APPROVED** [review]
> Just a suggestion, I would split the commit in 2 part, one with the rearrangements of the variables and a second wit the new variables for the sidecar sha image. It is just easier to check the new variable.

+1


Overall, the code looks good, and the tests are green.

File: tools/csv-generator/csv-generator.go [hunk 5]
```diff
@@ -63,7 +64,8 @@ func main() {
 	virtExportProxyImage := flag.String("virt-export-proxy-image", "", "custom image for virt-export-proxy. "+customImageExample)
 	virtExportServerImage := flag.String("virt-export-server-image", "", "custom image for virt-export-server. "+customImageExample)
 	gsImage := flag.String("gs-image", "", "custom image for gs. "+customImageExample)
-	prHelperImage := flag.String("pr-helper-image", "", "custom image for gs. "+customImageExample)
+	prHelperImage := flag.String("pr-helper-image", "", "custom image for pr-helper. "+customImageExample)
```
Comment (victortoso) [inline tools/csv-generator/csv-generator.go:67]: nit: this fix should probably be on its own commit

**Comment by victortoso** [comment]
minor nit only, /lgtm

**Comment by alicefr** [comment]
/approve
/hold

Hold to let you fix the last comments, but the code looks good.

File: tools/csv-generator/csv-generator.go (hunk 5 above)
Comment (dharmit) [reply tools/csv-generator/csv-generator.go:67]: Done. PTAL.

----
If you don't mind, how would you go about doing this? I'll share how I did it.
1. `git rebsae -i HEAD~4`
2. `edit` the specific commit - 6e083f4345 in this case.
3. Modify the line in question to what it was earlier.
4. Amend the commit with `git add . && git commit --amend`, and finish rebase with `git rebsae --continue`.
5. Make a new commit - b02c45b5 in this case.

I was hoping to be able to use `git reset` to do this instead, but couldn't figure it out.

Comment (victortoso) [reply tools/csv-generator/csv-generator.go:67]: For removing something from a commit and moving it to another, I usually do the following:
1. `git rebase -i ...`
2. `edit` the specific commit
3. `undo` the commit
4. open and keep your editor open!  remove what you don't want to see
5. git add the changed buffers and `redo` the commit
6. on the editor, `ctrl-z` to get the changes back
7. git add && git commit the new commit
8. finish rebase

Now, the `undo` and `redo` are actually git alias that I have here, they are:
```
[alias]
    undo = reset --soft HEAD~
    redo = commit -c ORIG_HEAD
```

When you do `redo`, you get back the commit log from the commit you did `undo`

Pretty sure there are plenty ways to go around this, the important part is being able to do it :smile: 
Thanks again!

**Comment by victortoso** [comment]
/lgtm
:rocket:

**Comment by vasiliy-ul** [comment]
@dharmit, there are compile failures.

/lgtm cancel

**Comment by dharmit** [comment]
> @dharmit, there are compile failures.
> 
> /lgtm cancel

Oops. My apologies. I didn't run `make test` or `make lint` before pushing. I have done that this time.

Can you please help trigger a couple of selected tests to verify if we can move further with full tests?

**Comment by vasiliy-ul** [comment]
I think the failure is not related to this PR. Probably the job is not very stable. But lets just try one more time.

**Comment by brianmcarey** [comment]
> > /test pull-kubevirt-goveralls
> 
> I wonder if it's a real failure. Let's see. 😕

Theres a known issue with the goveralls lane on main at the moment - @fossedihelm is working on a fix. The lane is not required at the moment so it should be ok.

**Comment by vasiliy-ul** [comment]
Ok, then

/lgtm

**Comment by dharmit** [comment]
@vasiliy-ul @brianmcarey should the 'goveralls' lane be manually overwritten? If yes, someone will have to do it on this PR.

**Comment by victortoso** [comment]
> should the 'goveralls' lane be manually overwritten

I think the job is not `required` so when tide finished the retesting, this should be merged, no?

**Comment by brianmcarey** [comment]
> > should the 'goveralls' lane be manually overwritten
> 
> I think the job is not `required` so when tide finish the retesting, this should be merged, no?

Yes tide should handle it - its in a pool waiting for retest. https://prow.ci.kubevirt.io/tide
